scRNA-seq/04-dimension_reduction_scRNA.Rmd

```diff
@@ -31,226 +42,294 @@ library(magrittr)
+We can do that with the `ncomponents` argument.
 
-```{r load_normalized}
-tm_normalized <- readr::read_rds(tm_normalized_file)
+```{r plotPCA34, live = TRUE}
+# plot PC3 and PC4
+plotReducedDim(normalized_sce, "PCA", ncomponents = c(3,4))
 ```
 
-How many cells are in this data? How many samples are there? How many cells from each tissue?
+### Modeling variance
 
-```{r explore_sce, live = TRUE}
-# explore summaries of the SCE data
-dim(tm_normalized) # column count is the number of cells
+The variation in gene expression we see among cells comes from a combination of variation due to technical effects and the biology we really care about. 
+In order to roughly account for this we could just take the largest variance genes, on the assumption that low variance genes are mostly just noise. 
+This is the default approach that `runPCA()` and `calculatePCA()` take, using 500 genes with the 
+greatest variance across cells to calculate the PCA matrix. 
+This approach also makes the PCA calculations somewhat faster.
 
-unique(tm_normalized$sample_id) # what are the samples
+If we want to be a bit more careful about it, we can model the variance in expression of each gene as a function of the mean expression for that gene.
+This is useful because we generally expect the variance to increase as mean expression increases, even if there is no biological signal in the expression variation.
 
-table(tm_normalized$tissue) # table of cells per tissue
+We will do this modeling of variance by expression with the `scran::modelGeneVar()` function, saving the results to a new variable. 
 
-as.data.frame(colData(tm_normalized)) %>% # dplyr summary table
-  dplyr::group_by(tissue) %>%
-  dplyr::summarize(n = dplyr::n(), 
-                   mean_detected = mean(detected), 
-                   max_detected = max(detected))
+```{r model_variance}
+gene_variance <- scran::modelGeneVar(normalized_sce)
 ```
-### PCA baseline
 
-Before we move on to UMAP, let's calculate the principle components for this full data set, and see what it can show us.
+Now let's plot the relationship between gene expression and variance we were discussing.
+Here we will also add the fitting curve that `scran::modelGeneVar()` created, which is stored as function in the  `$trend` slot of the `gene_variance` object.
+We can add a function like that curve to a `ggplot` with a `stat_function` layer.
+
+```{r plot_variance}
+ggplot(as.data.frame(gene_variance), aes(x = mean, y = total)) +
+  geom_point(alpha = 0.1) +
+  stat_function(fun = metadata(gene_variance)$trend, color = "blue") + 
+  xlab("Mean log-expression") + 
+  ylab("Variance") +
+  theme_bw()
+```
 
-In the chunk below, we calculate the PCs and plot the first two, colored by tissue, much as we did before.
+Now we can use `scran::getTopHVGs()` to select the genes that have the most biological variation (according to the model) and recalculate PCA scores using only those genes. 
```

Comment:
I feel pretty certain that we've already had this conversation, but can you remind me (roughly) how genes with biological variation are defined by this particular method? I can understand why you might not want to get into it here (again, we've probably had this exact conversation before!) but it is helpful for instructors to understand.

Reply:
I _believe_ (I checked, [I seem to be correct](https://rdrr.io/bioc/scran/man/modelGeneVar.html)) it models the gene variance relationship (shown in the plot) with a loess model or similar, then uses the residuals as the measure of biological variation. 

```suggestion
Now we can use `scran::getTopHVGs()` to select the genes that have the most biological variation (according to the model) and recalculate PCA scores using only those genes. 
(In practice, we are selecting the genes with the largest residual variation after removing technical variation modeled by the mean/variance relationship.)
```

Something like that?